lib/ts/recipe/session/sessionFunctions.ts

```diff
@@ -83,64 +85,85 @@ export async function getSession(
 }> {
     let handShakeInfo = await recipeImplementation.getHandshakeInfo();
     let accessTokenInfo;
-    /**
-     * if jwtSigningPublicKeyExpiryTime is expired, we call core
-     */
-    if (handShakeInfo.jwtSigningPublicKeyExpiryTime > Date.now()) {
-        try {
-            /**
-             * get access token info using existing signingKey
-             */
-            accessTokenInfo = await getInfoFromAccessToken(
-                accessToken,
-                handShakeInfo.jwtSigningPublicKey,
-                handShakeInfo.antiCsrf === "VIA_TOKEN" && doAntiCsrfCheck
-            );
-        } catch (err) {
-            /**
-             * if error type is not TRY_REFRESH_TOKEN, we return the
-             * error to the user
-             */
-            if (err.type !== STError.TRY_REFRESH_TOKEN) {
-                throw err;
-            }
-            /**
-             * if it comes here, it means token verification has failed.
-             * It may be due to:
-             *  - signing key was updated and this token was signed with new key
-             *  - access token is actually expired
-             *  - access token was signed with the older signing key
-             *
-             * if access token is actually expired, we don't need to call core and
-             * just return TRY_REFRESH_TOKEN to the client
-             *
-             * if access token creation time is after the signing key was last
-             * updated, we need to call core as there are chances that the token
-             * was signed with the updated signing key
-             *
-             * if access token creation time is before the signing key was last
-             * updated, we just return TRY_REFRESH_TOKEN to the client
-             */
-            let payload;
+
+    // If we have no key old enough to verify this access token we should reject it without calling the core
+    let foundOlderKey = false;
+    for (const key of handShakeInfo.jwtSigningPublicKeyList) {
+        // If the key is already expired we shouldn't use it for verification
+        if (key.expiryTime > Date.now()) {
```

Comment:
- We would expect the core to give only those tokens have not expired?
- Also, this check is not done in the core [here](https://github.com/supertokens/supertokens-core/pull/307/files#diff-5621d130c254985cb161a1060a401fcc48c5aadaf882e0cad1d52746eecb9d48R56). Is that ok in the core?

Reply:
Never access `handShakeInfo.jwtSigningPublicKeyList` directly and instead call a function which filters array to remove expired keys, and use that function everywhere. Therefore, you won't need to check `if (key.expiryTime > Date.now()) {`